measure_optimize.py

```diff
@@ -2,180 +2,2412 @@
+                    + str(self.n_total_measurements)
+                )
+
+            # check the time correlation matrice shape for each DCM
+            # loop over the index of DCM to retrieve the number of time points for DCM
+            for i in range(self.n_total_measurements):
+                # check row number
+                if len(error_cov[0]) != self.sens_info.Nt:
+                    raise ValueError(
+                        "error_cov[i] must have the shape Nt*Nt. Expect number of rows:"
+                        + str(self.sens_info.Nt)
+                    )
+                # check column number
+                if len(error_cov[0][0]) != self.sens_info.Nt:
+                    raise ValueError(
+                        "error_cov[i] must have the shape Nt*Nt. Expect number of columns:"
+                        + str(self.sens_info.Nt)
+                    )
+
+        elif error_option == CovarianceStructure.measure_correlation:
+            # check row number
+            if len(error_cov) != self.sens_info.total_measure_idx:
+                raise ValueError(
+                    "error_cov must have the same length as total_measure_idx. Expect number of rows:"
+                    + str(self.sens_info.total_measure_idx)
+                )
+            # check column number
+            if len(error_cov[0]) != self.sens_info.total_measure_idx:
+                raise ValueError(
+                    "error_cov[i] must have the same length as total_measure_idx. Expect number of columns:"
+                    + str(self.sens_info.total_measure_idx)
+                )
+
+        elif error_option == CovarianceStructure.time_measure_correlation:
+            # check row number
+            if len(error_cov) != self.total_num_time:
+                raise ValueError(
+                    "error_cov must have the shape total_num_time*total_num_time. Expect number of rows:"
+                    + str(self.total_num_time)
+                )
+            # check column number
+            if len(error_cov[0]) != self.total_num_time:
+                raise ValueError(
+                    "error_cov must have the shape total_num_time*total_num_time. Expect number of columns:"
+                    + str(self.total_num_time)
+                )
+
+    def _dynamic_flatten(self, jac):
+        """Update dynamic flattened matrix index.
+        Arguments
+        ---------
+        :param jac: Jacobian information for main class use, a 2D array of shape [N_total_meausrements * Np]
+
+        Returns
+        -------
+        dynamic_flatten matrix: flatten dynamic-cost measurements, not flatten static-costs, [s1, d1|t1, ..., d1|tN, s2]
+        Flatten matrix: flatten dynamic-cost and static-cost measuremenets
+        """
+
+        ### dynamic_flatten: to be decision matrix
+        jac_dynamic_flatten = []
+        # position index in jac_dynamic_flatten where each measurement starts
+        self.head_pos_dynamic_flatten = {}
```

Comment:
Why is this a dictionary? What are the keys and values? What does "head" mean in the variable name?

Reply:
all head --> first in variable names

Reply:
changes all head in variable names to first

Reply:
added keys and values comments